Native: `str(obj)` ignores `__str__` (returns garbage) — blocks `uuid` + object stringification

## Summary

On the native (`na`) pathway, `str(obj)` on a user object does **not** dispatch to the object's `__str__` method — it returns garbage (an unrendered pointer/value) instead of the string. There is no `__str__` lowering, so the canonical Python stringification protocol is unavailable on native.

Discovered while building the bundled native `uuid` module in #7002 (Phase 1 of #6978); it is the blocker that forced `uuid` to be deferred.

## Repro

```jac
obj P {
    has v: str;
    def __str__ -> str { return self.v; }
}

with entry {
    p = P(v="hello");
    print(str(p));   # expected: hello
}
```

`jac nacompile`d (and run): prints garbage (e.g. `�@`), not `hello`. (Same on the JIT path.)

- **Expected:** `str(p)` calls `P.__str__` → `hello`.
- **Actual:** `__str__` is ignored; `str()` of a struct pointer yields garbage.

## Impact

- **`uuid`** (#6978 Phase 1): `uuid.uuid4()`'s primary use is `str(uuid4())` → the `8-4-4-4-12` form. Without `__str__` that can't be rendered congruently, so `uuid` is deferred in #7002.
- **General:** `str(obj)`, `print(obj)`, and `f"{obj}"` for any user object all need `__str__` dispatch — broadly useful well beyond the stdlib (logging, debugging, CLIs).
- **Compounding factor:** a bundled `na_stdlib` module whose name matches a CPython stdlib module type-resolves to **CPython's typeshed stub**, so its public surface must match CPython's API exactly. A workaround method like `str_form()` is rejected by the type checker (`E1030: Type "UUID" has no attribute "str_form"`) because CPython's `uuid.UUID` has no such method — so `__str__` is the *only* congruent way to expose the string form. (This is also why `hashlib`'s `Hash.hexdigest()` worked: the names already matched CPython.)

## Related native gap (lesser; worked around)

`bytes` are immutable on native — subscript assignment is rejected:

```jac
with entry { b = b"\x00\x00"; b[0] = 65; }   # error[E5090]: Native pathway does not yet support subscript assignment 'b'
```

For `uuid` this meant the version/variant bits couldn't be set in place on the 16 random bytes (worked around by applying them at render time). Not the primary blocker, but a related `bytes`-ergonomics gap (`bytearray` / item-assignment) worth tracking. A constructor like `bytes([ints])` would also suffice for that case.

## Acceptance

- `str(obj)` / `print(obj)` / `f"{obj}"` dispatch to a user `__str__` method on both `nacompile` (AOT) and JIT.
- Falls back to a sensible default (e.g. `<Obj at 0x...>`) when no `__str__` is defined, matching today's behavior shape without garbage.
- Unblocks the bundled native `uuid` module (#6978 Phase 1).

Refs: #6978 (native stdlib roadmap), #7002 (`secrets`; `uuid` deferred here).

--- a/docs/docs/community/release_notes/unreleased/jaclang/7005.feature.md
+++ b/docs/docs/community/release_notes/unreleased/jaclang/7005.feature.md
@@ -1,0 +1,1 @@
+- **Native: `__str__` dispatch for `str()` / `print()` / f-strings**: On the native (`na`) pathway, `str(obj)`, `print(obj)`, and `f"{obj}"` now dispatch to a user object's `__str__` (resolving the runtime subclass through the vtable) instead of rendering the raw struct pointer as garbage, and fall back to a `<TypeName object at 0x...>` default when no `__str__` is defined.

--- a/jac/jaclang/compiler/passes/native/impl/primitives_native.impl.jac
+++ b/jac/jaclang/compiler/passes/native/impl/primitives_native.impl.jac
@@ -5858,6 +5858,13 @@ impl NativeBuiltinEmitter.emit_str(
     # pointer address.
     if isinstance(arg_val.type, ir.PointerType) {
         if arg_val.type != i8p {
+            # str(obj): dispatch to the user object's __str__ (or a default
+            # repr), instead of bitcasting the struct pointer to garbage (#7003).
+            # The builtin-arg release at the call site (calls.impl) owns `arg_val`.
+            obj_s = p._emit_obj_str(arg_val);
+            if obj_s is not None {
+                return obj_s;
+            }
             return b.bitcast(arg_val, i8p, name="str.identity");
         }
         # str(None): None lowers to a null i8*; CPython prints "None". Guard the

--- a/jac/jaclang/compiler/passes/native/na_ir_gen_pass.impl/builtins.impl.jac
+++ b/jac/jaclang/compiler/passes/native/na_ir_gen_pass.impl/builtins.impl.jac
@@ -75,6 +75,16 @@ impl NaIRGenPass._emit_print(nd: uni.FuncCall) -> (ir.Value | None) {
             fmt_parts.append("%s");
             args.append(self._emit_float_repr(val));
         } elif isinstance(val.type, ir.PointerType) {
+            # print(obj): stringify user objects via __str__ (#7003). The source
+            # object temp is consumed here; the produced string is owned and is
+            # released by the loop below alongside other owned temporaries.
+            obj_s = self._emit_obj_str(val);
+            if obj_s is not None {
+                if self._is_owned(val) {
+                    self._emit_rc_release_typed(val, op="print_obj_src", loc="");
+                }
+                val = obj_s;
+            }
             fmt_parts.append("%s");
             args.append(val);
         }
@@ -232,6 +242,9 @@ impl NaIRGenPass._codegen_string(nd: uni.String) -> (ir.Value | None) {
 impl NaIRGenPass._codegen_fstring(nd: uni.FString) -> (ir.Value | None) {
     fmt_parts: list[str] = [];
     args: list[ir.Value] = [];
+    # Strings produced by object __str__ dispatch (#7003): owned temporaries that
+    # must be released once the snprintf below has consumed them.
+    obj_strs: list[ir.Value] = [];
     for part in nd.parts {
         if isinstance(part, uni.String) {
             # Literal text portion — `lit_value` decodes escapes the same way
@@ -257,6 +270,19 @@ impl NaIRGenPass._codegen_fstring(nd: uni.FString) -> (ir.Value | None) {
                 fmt_parts.append("%s");
                 args.append(self._emit_float_repr(val));
             } elif isinstance(val.type, ir.PointerType) {
+                # f"{obj}": stringify user objects via __str__ (#7003). Consume
+                # the source object temp now; the produced string is owned and is
+                # released after the snprintf that reads it.
+                obj_s = self._emit_obj_str(val);
+                if obj_s is not None {
+                    if self._is_owned(val) {
+                        self._emit_rc_release_typed(
+                            val, op="fstr_obj_src", loc=""
+                        );
+                    }
+                    val = obj_s;
+                    obj_strs.append(obj_s);
+                }
                 fmt_parts.append("%s");
                 args.append(val);
             }
@@ -285,6 +311,11 @@ impl NaIRGenPass._codegen_fstring(nd: uni.FString) -> (ir.Value | None) {
         call_args.append(a);
     }
     self.builder.call(snprintf_fn, call_args, name="fstr.len");
+    # RC: release the per-object __str__ strings now that snprintf copied them
+    # into the f-string buffer (#7003).
+    for _obj_s in obj_strs {
+        self._emit_rc_release_typed(_obj_s, op="fstr_obj_tmp", loc="");
+    }
     # RC: caller owns this freshly-allocated fstring buffer (rc=1 from malloc).
     # Mark it so the assign handler skips the extra RETAIN (avoiding rc=2 leak).
     self._mark_owned(buf);

--- a/jac/jaclang/compiler/passes/native/na_ir_gen_pass.impl/builtins.impl.jac
+++ b/jac/jaclang/compiler/passes/native/na_ir_gen_pass.impl/builtins.impl.jac
@@ -276,9 +276,7 @@ impl NaIRGenPass._codegen_fstring(nd: uni.FString) -> (ir.Value | None) {
                 obj_s = self._emit_obj_str(val);
                 if obj_s is not None {
                     if self._is_owned(val) {
-                        self._emit_rc_release_typed(
-                            val, op="fstr_obj_src", loc=""
-                        );
+                        self._emit_rc_release_typed(val, op="fstr_obj_src", loc="");
                     }
                     val = obj_s;
                     obj_strs.append(obj_s);

--- a/jac/jaclang/compiler/passes/native/na_ir_gen_pass.impl/objects.impl.jac
+++ b/jac/jaclang/compiler/passes/native/na_ir_gen_pass.impl/objects.impl.jac
@@ -1959,28 +1959,63 @@ impl NaIRGenPass._emit_obj_str(val: ir.Value) -> (ir.Value | None) {
             break;
         }
     }
+    # Null guard: a null struct pointer is the native form of a `T | None` value
+    # that is None. Dispatching would deref-crash on the vtable GEP (as
+    # _codegen_method_call guards too), so short-circuit to "None" -- matching
+    # str(None) / print(None) / f"{None}" on the sv pathway.
+    i8p = ir.IntType(8).as_pointer();
+    is_null = self.builder.icmp_unsigned(
+        "==", val, ir.Constant(val.type, None), name="objstr.isnull"
+    );
+    fn = self.builder.basic_block.function;
+    null_bb = fn.append_basic_block(name="objstr.null");
+    live_bb = fn.append_basic_block(name="objstr.live");
+    cont_bb = fn.append_basic_block(name="objstr.cont");
+    self.builder.cbranch(is_null, null_bb, live_bb);
+    # None path.
+    self.builder.position_at_end(null_bb);
+    none_s = self._make_global_string("None", name_prefix=".objstr.none");
+    null_end = self.builder.basic_block;
+    self.builder.branch(cont_bb);
+    # Live (non-null) path: dispatch __str__, or fall back to the default repr.
+    self.builder.position_at_end(live_bb);
     if func is None {
-        return self._emit_default_obj_repr(val, type_name);
+        live_res = self._emit_default_obj_repr(val, type_name);
+    } else {
+        # The receiver is borrowed: __str__ does not consume self, and the
+        # value's producer owns its release (str()'s builtin-arg release in
+        # calls.impl, or the print/f-string source-object release). Hide
+        # ownership across the dispatch so the vtable path's owned-arg release
+        # leaves the receiver alone, then restore it.
+        was_owned = self._is_owned(val);
+        if was_owned {
+            val.is_owned = False;
+        }
+        result = None;
+        # Virtual dispatch resolves the runtime subclass's __str__; falls back
+        # to a direct call for non-vtable (standalone) types.
+        if self._arch_has_vtable(type_name) {
+            result = self._codegen_virtual_call(val, "__str__", [val], type_name);
+        }
+        if result is None {
+            coerced = [self._coerce_type(val, func.args[0].type)];
+            result = self.builder.call(func, coerced, name="call.__str__");
+        }
+        if was_owned {
+            val.is_owned = True;
+        }
+        live_res = result;
     }
-    was_owned = self._is_owned(val);
-    if was_owned {
-        val.is_owned = False;
-    }
-    result = None;
-    # Virtual dispatch resolves the runtime subclass's __str__; falls back to a
-    # direct call for non-vtable (standalone) types.
-    if self._arch_has_vtable(type_name) {
-        result = self._codegen_virtual_call(val, "__str__", [val], type_name);
-    }
-    if result is None {
-        coerced = [self._coerce_type(val, func.args[0].type)];
-        result = self.builder.call(func, coerced, name="call.__str__");
-    }
-    if was_owned {
-        val.is_owned = True;
-    }
+    live_res = self._coerce_type(live_res, i8p);
+    live_end = self.builder.basic_block;
+    self.builder.branch(cont_bb);
+    # Merge the None and dispatch paths.
+    self.builder.position_at_end(cont_bb);
+    phi = self.builder.phi(i8p, name="objstr.result");
+    phi.add_incoming(none_s, null_end);
+    phi.add_incoming(live_res, live_end);
     # The produced string is a fresh rc=1 value the caller now owns.
-    return self._mark_owned(result);
+    return self._mark_owned(phi);
 }
 
 """Default object stringification when no `__str__` is defined (#7003).

--- a/jac/jaclang/compiler/passes/native/na_ir_gen_pass.impl/objects.impl.jac
+++ b/jac/jaclang/compiler/passes/native/na_ir_gen_pass.impl/objects.impl.jac
@@ -1931,6 +1931,93 @@ impl NaIRGenPass._codegen_method_call(nd: uni.FuncCall) -> (ir.Value | None) {
     return result;
 }
 
+"""Stringify a user-object value via its `__str__` method (#7003).
+
+Returns a freshly-owned `i8*` string when `val` is a user struct pointer:
+either the result of dispatching to a user `__str__` (along the MRO, through the
+vtable so a base-typed receiver resolves the runtime subclass's override), or a
+`<TypeName object at 0x...>` default when no `__str__` is defined. Returns None
+when `val` is not a user struct, so the callers (str()/print()/f-strings) keep
+their existing scalar/string handling.
+
+The receiver is *borrowed*: `__str__` does not consume `self`, and the value's
+producer owns its release (str()'s builtin-arg release in calls.impl, or the
+print/f-string source-object release). Ownership is hidden across the dispatch
+so the vtable path's owned-arg release leaves the receiver alone, then restored
+so that producer can still release it once."""
+impl NaIRGenPass._emit_obj_str(val: ir.Value) -> (ir.Value | None) {
+    type_name = self._infer_type_name(val);
+    if type_name is None {
+        return None;
+    }
+    # Resolve a user __str__ along the MRO (mirrors _codegen_method_call).
+    func = None;
+    for check_name in self._mro_of(type_name) {
+        cand = f"{check_name}.__str__";
+        if cand in self.method_funcs {
+            func = self.method_funcs[cand];
+            break;
+        }
+    }
+    if func is None {
+        return self._emit_default_obj_repr(val, type_name);
+    }
+    was_owned = self._is_owned(val);
+    if was_owned {
+        val.is_owned = False;
+    }
+    result = None;
+    # Virtual dispatch resolves the runtime subclass's __str__; falls back to a
+    # direct call for non-vtable (standalone) types.
+    if self._arch_has_vtable(type_name) {
+        result = self._codegen_virtual_call(val, "__str__", [val], type_name);
+    }
+    if result is None {
+        coerced = [self._coerce_type(val, func.args[0].type)];
+        result = self.builder.call(func, coerced, name="call.__str__");
+    }
+    if was_owned {
+        val.is_owned = True;
+    }
+    # The produced string is a fresh rc=1 value the caller now owns.
+    return self._mark_owned(result);
+}
+
+"""Default object stringification when no `__str__` is defined (#7003).
+
+Renders `<TypeName object at 0x...>` into a fresh RC-managed `str` buffer,
+mirroring CPython's default `object.__repr__` shape instead of leaking the raw
+struct pointer as garbage. The receiver is only read (ptrtoint), never consumed."""
+impl NaIRGenPass._emit_default_obj_repr(val: ir.Value, type_name: str) -> ir.Value {
+    b = self.builder;
+    i8p = ir.IntType(8).as_pointer();
+    i64 = ir.IntType(64);
+    # "<" + name + " object at 0x" + up to 16 hex digits + ">" + NUL; pad.
+    buf_len = len(type_name) + 48;
+    buf = self._rc_call_with_ctx(
+        self.rc_alloc_fn,
+        [ir.Constant(i64, buf_len)],
+        op="obj_repr",
+        loc="",
+        name="obj.repr.buf"
+    );
+    self._emit_set_type_tag(b, buf, "str");
+    snprintf = self._get_or_declare_extern(
+        "snprintf", ir.IntType(32), [i8p, i64, i8p], var_arg=True
+    );
+    fmt = self._make_global_string(
+        "<%s object at 0x%llx>", name_prefix=".obj.repr.fmt"
+    );
+    name_g = self._make_global_string(type_name, name_prefix=".obj.repr.name");
+    addr = b.ptrtoint(val, i64, name="obj.repr.addr");
+    b.call(
+        snprintf,
+        [buf, ir.Constant(i64, buf_len), fmt, name_g, addr],
+        name="obj.repr.snprintf"
+    );
+    return self._mark_owned(buf);
+}
+
 """Generate field access: GEP into struct + load (with AttributeError null check)."""
 impl NaIRGenPass._codegen_field_access(
     obj_val: ir.Value, field_name: str, type_name: str

--- a/jac/jaclang/compiler/passes/native/na_ir_gen_pass.jac
+++ b/jac/jaclang/compiler/passes/native/na_ir_gen_pass.jac
@@ -652,6 +652,9 @@ obj NaIRGenPass(ModuleCodegenPass) {
     def _is_walker_arch(name: str) -> bool;
     def _codegen_instantiation(nd: (uni.FuncCall | uni.Name)) -> (ir.Value | None);
     def _codegen_method_call(nd: uni.FuncCall) -> (ir.Value | None);
+    # Object stringification (#7003): dispatch str()/print()/f"{}" to __str__.
+    def _emit_obj_str(val: ir.Value) -> (ir.Value | None);
+    def _emit_default_obj_repr(val: ir.Value, type_name: str) -> ir.Value;
     def _codegen_field_access(
         obj_val: ir.Value, field_name: str, type_name: str
     ) -> (ir.Value | None);

--- a/jac/tests/compiler/passes/native/fixtures/obj_str.na.jac
+++ b/jac/tests/compiler/passes/native/fixtures/obj_str.na.jac
@@ -5,13 +5,18 @@ none is defined, and resolve the runtime subclass through the vtable.
 
 obj P {
     has v: str;
-    def __str__ -> str { return self.v; }
+
+    def __str__ -> str {
+        return self.v;
+    }
 }
 
 obj Q {
     has n: int;
     # __str__ that itself builds a string (nested f-string).
-    def __str__ -> str { return f"Q({self.n})"; }
+    def __str__ -> str {
+        return f"Q({self.n})";
+    }
 }
 
 # No __str__ anywhere -> sensible default repr, not a garbage pointer.
@@ -23,29 +28,34 @@ obj Plain {
 # __str__ through the vtable (mirrors the class_name.na.jac dispatch shape).
 obj Animal {
     has name: str;
-    def __str__ -> str { return f"Animal:{self.name}"; }
+
+    def __str__ -> str {
+        return f"Animal:{self.name}";
+    }
 }
 
 obj Dog(Animal) {
-    def __str__ -> str { return f"Dog:{self.name}"; }
+    def __str__ -> str {
+        return f"Dog:{self.name}";
+    }
 }
 
-def obj_str() -> str {
+def obj_str -> str {
     p = P(v="hello");
     return str(p);
 }
 
-def obj_str_computed() -> str {
+def obj_str_computed -> str {
     q = Q(n=42);
     return str(q);
 }
 
-def obj_in_fstring() -> str {
+def obj_in_fstring -> str {
     p = P(v="world");
     return f"[{p}]";
 }
 
-def obj_default_repr() -> str {
+def obj_default_repr -> str {
     return str(Plain(x=7));
 }
 
@@ -54,12 +64,22 @@ def animal_str(a: Animal) -> str {
     return str(a);
 }
 
-def poly_str() -> str {
+def poly_str -> str {
     return animal_str(Dog(name="rex"));
 }
 
-def base_str() -> str {
+def base_str -> str {
     return animal_str(Animal(name="generic"));
+}
+
+# A `P | None` value that is None lowers to a null struct pointer; str() must
+# render "None" rather than dereference it (vtable GEP) and crash.
+def maybe_str(flag: int) -> str {
+    p: P | None = None;
+    if flag {
+        p = P(v="set");
+    }
+    return str(p);
 }
 
 with entry {

--- a/jac/tests/compiler/passes/native/fixtures/obj_str.na.jac
+++ b/jac/tests/compiler/passes/native/fixtures/obj_str.na.jac
@@ -1,0 +1,67 @@
+"""Native __str__ dispatch (#7003): str(obj)/print(obj)/f"{obj}" must call the
+user object's __str__, fall back to a `<TypeName object at 0x...>` default when
+none is defined, and resolve the runtime subclass through the vtable.
+"""
+
+obj P {
+    has v: str;
+    def __str__ -> str { return self.v; }
+}
+
+obj Q {
+    has n: int;
+    # __str__ that itself builds a string (nested f-string).
+    def __str__ -> str { return f"Q({self.n})"; }
+}
+
+# No __str__ anywhere -> sensible default repr, not a garbage pointer.
+obj Plain {
+    has x: int;
+}
+
+# Polymorphism: a base-typed receiver must resolve the runtime subclass's
+# __str__ through the vtable (mirrors the class_name.na.jac dispatch shape).
+obj Animal {
+    has name: str;
+    def __str__ -> str { return f"Animal:{self.name}"; }
+}
+
+obj Dog(Animal) {
+    def __str__ -> str { return f"Dog:{self.name}"; }
+}
+
+def obj_str() -> str {
+    p = P(v="hello");
+    return str(p);
+}
+
+def obj_str_computed() -> str {
+    q = Q(n=42);
+    return str(q);
+}
+
+def obj_in_fstring() -> str {
+    p = P(v="world");
+    return f"[{p}]";
+}
+
+def obj_default_repr() -> str {
+    return str(Plain(x=7));
+}
+
+# `a` is statically Animal; str() lowers a virtual __str__ call.
+def animal_str(a: Animal) -> str {
+    return str(a);
+}
+
+def poly_str() -> str {
+    return animal_str(Dog(name="rex"));
+}
+
+def base_str() -> str {
+    return animal_str(Animal(name="generic"));
+}
+
+with entry {
+    print(P(v="printed"));
+}

--- a/jac/tests/compiler/passes/native/test_native_obj_str.jac
+++ b/jac/tests/compiler/passes/native/test_native_obj_str.jac
@@ -53,10 +53,12 @@ def capture_stdout(fn: object) -> bytes {
     } finally {
         os.dup2(saved, 1);
         os.close(saved);
+        # Close the write-end to signal EOF, then drain and close the read-end.
+        # Kept in `finally` so a raising fn() never leaks the pipe fds.
         os.close(w);
+        data = os.read(r, 65536);
+        os.close(r);
     }
-    data = os.read(r, 65536);
-    os.close(r);
     return data;
 }
 
@@ -100,6 +102,14 @@ test "str(obj) on a base instance uses the base __str__" {
     (eng, _) = compile_native("obj_str.na.jac");
     f = get_func(eng, "base_str", ctypes.c_char_p);
     assert f() == b"Animal:generic";
+}
+
+# --- null receiver (T | None) ---
+test "str(obj) on a null (None) struct pointer renders \"None\", not a crash" {
+    (eng, _) = compile_native("obj_str.na.jac");
+    f = get_func(eng, "maybe_str", ctypes.c_char_p, ctypes.c_int64);
+    assert f(1) == b"set";
+    assert f(0) == b"None";
 }
 
 # --- print(obj) ---

--- a/jac/tests/compiler/passes/native/test_native_obj_str.jac
+++ b/jac/tests/compiler/passes/native/test_native_obj_str.jac
@@ -1,0 +1,110 @@
+"""End-to-end tests for native object stringification (#7003).
+
+`str(obj)`, `f"{obj}"`, and `print(obj)` must dispatch to a user object's
+`__str__` method (resolving the runtime subclass through the vtable), and fall
+back to a `<TypeName object at 0x...>` default -- not a garbage pointer -- when
+no `__str__` is defined. All assertions hold the native (`na`) result to the
+shape it would take on the `sv` pathway.
+"""
+
+import ctypes;
+import os;
+import from tests.support { TESTS_DIR }
+import from jaclang.jac0core.program { JacProgram }
+import jaclang.jac0core.unitree as uni;
+import from jaclang.compiler.passes.native.llvm.binding { ExecutionEngine }
+
+glob FIXTURES = os.path.join(
+         TESTS_DIR, "compiler", "passes", "native", "fixtures"
+     );
+
+"""Compile a .na.jac fixture and return the JIT engine."""
+def compile_native(fixture: str) -> tuple[ExecutionEngine, uni.Module] {
+    prog = JacProgram();
+    ir = prog.compile(file_path=str(os.path.join(FIXTURES, fixture)));
+    errors = [str(e) for e in prog.errors_had] if prog.errors_had else [];
+    assert not prog.errors_had , f"Compilation errors in {fixture}: {errors}";
+    eng = ir.gen.native_engine;
+    assert eng is not None , f"No native engine produced for {fixture}";
+    return (eng, ir);
+}
+
+"""Get a ctypes-callable function from the JIT engine."""
+def get_func(eng: object, name: str, restype: type | None, *argtypes: type) -> object {
+    addr = eng.get_function_address(name);
+    assert addr != 0 , f"Function '{name}' not found in JIT engine";
+    functype = ctypes.CFUNCTYPE(restype, *argtypes);
+    return functype(addr);
+}
+
+"""Run a void JIT function and capture what it writes to fd 1 (C stdout).
+
+`contextlib.redirect_stdout` only sees Python writes; the JIT's printf goes
+straight to the C runtime, so we redirect the OS-level file descriptor and
+flush libc's stdio buffer before restoring it."""
+def capture_stdout(fn: object) -> bytes {
+    libc = ctypes.CDLL(None);
+    (r, w) = os.pipe();
+    saved = os.dup(1);
+    os.dup2(w, 1);
+    try {
+        fn();
+        libc.fflush(None);
+    } finally {
+        os.dup2(saved, 1);
+        os.close(saved);
+        os.close(w);
+    }
+    data = os.read(r, 65536);
+    os.close(r);
+    return data;
+}
+
+# --- str(obj) ---
+test "str(obj) dispatches to __str__" {
+    (eng, _) = compile_native("obj_str.na.jac");
+    f = get_func(eng, "obj_str", ctypes.c_char_p);
+    assert f() == b"hello";
+}
+
+test "str(obj) with a computed (f-string) __str__" {
+    (eng, _) = compile_native("obj_str.na.jac");
+    f = get_func(eng, "obj_str_computed", ctypes.c_char_p);
+    assert f() == b"Q(42)";
+}
+
+# --- f"{obj}" ---
+test "f-string interpolates obj via __str__" {
+    (eng, _) = compile_native("obj_str.na.jac");
+    f = get_func(eng, "obj_in_fstring", ctypes.c_char_p);
+    assert f() == b"[world]";
+}
+
+# --- default repr fallback (no __str__) ---
+test "str(obj) without __str__ falls back to default repr, not garbage" {
+    (eng, _) = compile_native("obj_str.na.jac");
+    f = get_func(eng, "obj_default_repr", ctypes.c_char_p);
+    got = f();
+    assert got.startswith(b"<Plain object at 0x") , f"got: {got!r}";
+    assert got.endswith(b">") , f"got: {got!r}";
+}
+
+# --- polymorphism (vtable dispatch through a base-typed receiver) ---
+test "str(obj) resolves the runtime subclass __str__" {
+    (eng, _) = compile_native("obj_str.na.jac");
+    f = get_func(eng, "poly_str", ctypes.c_char_p);
+    assert f() == b"Dog:rex";
+}
+
+test "str(obj) on a base instance uses the base __str__" {
+    (eng, _) = compile_native("obj_str.na.jac");
+    f = get_func(eng, "base_str", ctypes.c_char_p);
+    assert f() == b"Animal:generic";
+}
+
+# --- print(obj) ---
+test "print(obj) writes __str__ to stdout" {
+    (eng, _) = compile_native("obj_str.na.jac");
+    entry = get_func(eng, "jac_entry", None);
+    assert capture_stdout(entry) == b"printed\n";
+}

